packages/core/src/upload-files/tar.ts

```diff
@@ -0,0 +1,139 @@
+import util from 'util'
+import path from 'path'
+import fs from 'fs'
+import { pack, Headers, Pack } from 'tar-stream'
+import { Writable, pipeline } from 'stream'
+import { TransferProgressEmitter, transferProgressEmitter } from './progress'
+import { FileInfo, FileToCopy } from './files'
+import { Visitor, fsWalker } from './walk'
+
+const headerFromStats = (local: FileInfo, remote: string): Headers | undefined => {
+  const { stats, symlinkTarget } = local
+  const header: Headers = {
+    name: remote,
+    mode: stats.mode,
+    mtime: stats.mtime,
+    size: stats.size,
+    uid: stats.uid,
+    gid: stats.gid,
+    type: 'file',
+  }
+
+  if (stats.isDirectory()) {
+    header.type = 'directory'
+  } else if (stats.isSymbolicLink()) {
+    header.type = 'symlink'
+    header.linkname = symlinkTarget
+    header.size = 0
+  } else if (stats.isFIFO()) {
+    header.type = 'fifo'
+  } else if (stats.isSocket()) {
+    return undefined
+  }
+
+  return header
+}
+
+const addEntry = (
+  p: Pack,
+  emitter: TransferProgressEmitter,
+  header: Headers,
+  filePath: string,
+) => new Promise<void>((resolve, reject) => {
+  const stream = p.entry(header, err => {
+    if (err) {
+      p.destroy(err)
+      reject(err)
+      return
+    }
+    resolve()
+  })
+
+  if (!stream) {
+    process.stderr.write(`no stream: ${filePath}, ${util.inspect(stream)}, ${util.inspect(p)} ${util.inspect(header)}\n`)
+    resolve()
+    return
+  }
+
+  if (header.type === 'file') {
+    emitter.emit('file', filePath)
+
+    const rs = fs.createReadStream(filePath)
+    rs.on('data', chunk => { emitter.emit('bytes', { bytes: chunk.length, file: filePath }) })
+    pipeline(rs, stream, err => {
+      if (err) {
+        p.destroy(err)
+        reject(err)
+      }
+    })
+  }
+})
+
+const chainPromise = () => {
+  let pr = Promise.resolve()
+  return {
+    chain: (f: () => Promise<void>) => { pr = pr.then(f) },
+    promise: () => pr,
+  }
+}
+
+export const tarStream = (filesToCopy: FileToCopy[] = []) => {
```

Comment:
convenience, instead of constructing then mutating, you can do `tarStream(file).finalize(...)`

Reply:
It can cause unexpected desired effect when passing an array to a mutating function without it being explicit.
It can be solved easily by copying the array, replace initialization with add(...files), or use a second array for the addition and iterate both.

Reply:
> It can cause unexpected desired effect when passing an array to a mutating function without it being explicit. It can be solved easily by copying the array, replace initialization with add(...files), or use a second array for the addition and iterate both.

fixed